adding profiles.json documentation (#883)
microsoft/terminal: doc/ProfilesJSONDocumentation.md

# Profiles.json Documentation

| Property | Type | Description |
| -------- | ---- | ----------- |
| **Global** |  | **Properties listed below affect the entire window, regardless of the property settings.** |

> | alwaysShowTabs | Boolean | When set to `true`, tabs are always displayed. When set to `false` and `experimental_showTabsInTitlebar` is set to `false`, tabs only appear after typing <kbd>Ctrl</kbd> + <kbd>T</kbd>. |

Reviewer: since these are key names, should we enclose them in literal backticks as well?

| defaultProfile | String | Sets the default profile. Opens by typing <kbd>Ctrl</kbd> + <kbd>T</kbd> or by clicking the '+' icon. The guid of the desired default profile is used as the value. |
| experimental_showTabsInTitlebar | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |
| initalCols | Integer | The number of columns displayed in the window upon first load. |
| initialRows | Integer | The number of rows displayed in the window upon first load. |
| requestedTheme | String | Sets the theme of the tab bar. Possible values: `"light"`, `"dark"`, `"system"` |
| showTerminalTitleInTitlebar | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
| **profiles** | **Array[Object]** | **Properties listed below are specific to each unique profile.** |

> | acrylicOpacity | Integer | Sets the transparency of the window for the profile. Accepts values from 0-1. |

Reviewer 2: We should probably also call out that this is only used when `useAcrylic` is `true`

> | background | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. |

Reviewer 2: Should we call out that the colors should be a hex string, i.e. `#rrggbb`?

| closeOnExit | Boolean | When set to `true`, the selected tab closes when `exit` is typed. When set to `false`, the tab will remain open when `exit` is typed. |

> | colorscheme | String | Name of the color scheme used in the profile that is defined under `schemes`. |

Reviewer 2: ```suggestion
| colorscheme | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |
```

> | colorTable | Array[String] | Array of colors used in the profile if `colorscheme` is not set. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |

Reviewer 2: Did we change these to be named as well as opposed to the raw table?

| commandline | String | Executable used in the profile. |
| cursorColor | String | Sets the cursor color for the profile. |
| cursorHeight | Integer | Sets the height of the cursor. Only works when `cursorShape` is set to `"vintage"`. Accepts values from 25-100. |
| cursorShape | String | Sets the cursor shape for the profile. Possible values: `"vintage"`, `"bar"`, `"underscore"`, `"filledBox"`, `"emptyBox"` |
| fontFace | String | Name of the font face used in the profile. |
| fontSize | Integer | Sets the font size. |
| foreground | String | Sets the foreground color of the profile. Overrides `foreground` set in color scheme if `colorscheme` is set. |

> | guid | String | Unique identifier of the profile. |

Reviewer: Hmmm. Do we need some indicator about what properties are required? perhaps we should even sort them, required first?

the `guid` is utterly required. `colorTable` isn't.
for a color scheme, the name is required.

Maybe this makes sense:

| Prop | Req | Type | Doc |
| - | - | - | - |
| `guid` | *required* | GUID String | etc
| `colorScheme` | | string | color scheme name

where `guid` unequivocally comes first as it's like the "primary key" of the type

Reviewer 3: How about making required properties bold and insert an empty line before Arrays (profiles, schemes) to emphasize that these are array objects with their properties below.

Reviewer 2: I agree that we need to make the optional properties clearly distinct from the required ones. I'm a little worried that bold might not be enough to differentiate, since the **globals**, **profiles**, etc, headers are all bolded was well.

> | guid | String | Unique identifier of the profile. |

Reviewer: for guid specifically we might need to indicate that it's a _guid_ in registry format (`{00000000-0000-0000-0000-000000000000}`)

| historySize | Integer | The number of lines above the ones displayed in the window you can scroll back to. |
| icon | String | Image file location of the icon used in the profile. Displays within the tab and the dropdown menu. |
| name | String | Name of the profile. Displays in the dropdown menu. |
| padding | String | Sets the padding around the text within the window. Can have three different formats: `"#"` sets the same padding for all sides, `"#, #"` sets the same padding for top-bottom and right-left, and `"#, #, #, #"` sets the padding individually for top, right, bottom, and left. |
| scrollbarState | String | Defines the visibility of the scrollbar. Possible values: `"visible"`, `"hidden"` |
| snapOnInput | Boolean | When set to `true`, the window will scroll to the command input line when typing. When set to `false`, the window will not scroll when you start typing. |
| startingDirectory | String | The directory the shell starts in when it is loaded. |
| useAcrylic | Boolean | When set to `true`, the window will have an acrylic background. When set to `false`, the window will have a plain, untextured background. |

> | **schemes** | **Array[Object]** | **Properties listed below are specific to each color theme.** |

Reviewer 2: ```suggestion
| **schemes** | **Array[Object]** | **Properties listed below are specific to each color scheme.** |
```
I want to make sure we keep `schemes` separate from `themes`, because when we do add support for theming, it's going to be a very different thing than just the color schemes :)

| background | String | Sets the background color of the color scheme. |

> | colors | Array[String] | Array of colors used in the color scheme. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |

Reviewer 2: This was also changed, see #742

| `alwaysShowTabs` | _Required_ | Boolean | When set to `true`, tabs are always displayed. When set to `false` and `experimental_showTabsInTitlebar` is set to `false`, tabs only appear after typing <kbd>Ctrl</kbd> + <kbd>T</kbd>. |
| `defaultProfile` | _Required_ | String | Sets the default profile. Opens by typing <kbd>Ctrl</kbd> + <kbd>T</kbd> or by clicking the '+' icon. The guid of the desired default profile is used as the value. |
| `initialCols` | _Required_ | Integer | The number of columns displayed in the window upon first load. |
| `initialRows` | _Required_ | Integer | The number of rows displayed in the window upon first load. |
| `requestedTheme` | _Required_ | String | Sets the theme of the application. Possible values: `"light"`, `"dark"`, `"system"` |
| `showTerminalTitleInTitlebar` | _Required_ | Boolean | When set to `true`, titlebar displays the title of the selected tab. When set to `false`, titlebar displays "Windows Terminal". |
| `experimental_showTabsInTitlebar` | Optional | Boolean | When set to `true`, the tabs are moved into the titlebar and the titlebar disappears. When set to `false`, the titlebar sits above the tabs. |

> | **profiles** | | **Array[Object]** | **Properties listed below are specific to each unique profile.** |

Reviewer 4: Global above was capitalized. profiles, schemes, and keybindings are not... intentional?

Author: Intentional - profiles, schemes, and keybindings are lowercase in the profiles.json file whereas Global isn't included in the file.

Reviewer 4: Ah OK.

> | `acrylicOpacity` | _Required_ | Number | When `useAcrylic` is set to `true`, it sets the transparency of the window for the profile. Accepts values from 0-1. |

Reviewer 4: Accepts... floating point values from 0-1?

| `background` | _Required_ | String | Sets the background color of the profile. Overrides `background` set in color scheme if `colorscheme` is set. Uses hex color format: `"#rrggbb"`. |
| `closeOnExit` | _Required_ | Boolean | When set to `true`, the selected tab closes when `exit` is typed. When set to `false`, the tab will remain open when `exit` is typed. |

> | `colorscheme` | _Required_ | String | Name of the terminal color scheme to use. Color schemes are defined under `schemes`. |

Reviewer 4: There's an inflight PR to change this to `colorScheme`... #1054 (and below reference)

| `commandline` | _Required_ | String | Executable used in the profile. |
| `cursorColor` | _Required_ | String | Sets the cursor color for the profile. Uses hex color format: `"#rrggbb"`. |

> | `cursorShape` | _Required_ | String | Sets the cursor shape for the profile. Possible values: `"vintage"`, `"bar"`, `"underscore"`, `"filledBox"`, `"emptyBox"` |

Reviewer 4: In this case (and the other cases where we say what the string options are), should we provide an example of what each one does either as a description or as a screenshot?

| `fontFace` | _Required_ | String | Name of the font face used in the profile. |
| `fontSize` | _Required_ | Integer | Sets the font size. |
| `guid` | _Required_ | String | Unique identifier of the profile. Written in registry format: `"{00000000-0000-0000-0000-000000000000}"`. |
| `historySize` | _Required_ | Integer | The number of lines above the ones displayed in the window you can scroll back to. |
| `name` | _Required_ | String | Name of the profile. Displays in the dropdown menu. |

> | `padding` | _Required_ | String | Sets the padding around the text within the window. Can have three different formats: `"#"` sets the same padding for all sides, `"#, #"` sets the same padding for top-bottom and right-left, and `"#, #, #, #"` sets the padding individually for top, right, bottom, and left. |

Reviewer 4: I am pretty sure that the order is "top, left, bottom, right". I've never seen a rectangle specified with right before left...

Reviewer 4: Same goes for top-bottom and left-right...

Author: I used this for reference: https://www.w3schools.com/cssref/pr_padding.asp

Reviewer 4: OK, but this isn't CSS.

This is the code calling for the thicknesses to be set up (for padding):
https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalControl/TermControl.cpp#L1338

And here's the function specification:
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.thicknesshelper.fromlengths#Windows_UI_Xaml_ThicknessHelper_FromLengths_System_Double_System_Double_System_Double_System_Double_

Which actually means we're both wrong. It's Left, Top, Right, Bottom.

| `snapOnInput` | _Required_ | Boolean | When set to `true`, the window will scroll to the command input line when typing. When set to `false`, the window will not scroll when you start typing. |
| `startingDirectory` | _Required_ | String | The directory the shell starts in when it is loaded. |
| `useAcrylic` | _Required_ | Boolean | When set to `true`, the window will have an acrylic background. When set to `false`, the window will have a plain, untextured background. |
| `colorTable` | Optional | Array[String] | Array of colors used in the profile if `colorscheme` is not set. Colors use hex color format: `"#rrggbb"`. Ordering is as follows: `[black, red, green, yellow, blue, magenta, cyan, white, bright black, bright red, bright green, bright yellow, bright blue, bright magenta, bright cyan, bright white]` |

> | `cursorHeight` | Optional | Integer | Sets the height of the cursor. Only works when `cursorShape` is set to `"vintage"`. Accepts values from 25-100. |

Reviewer 4: What does 25-100 mean? Clarify.

(It's the percentage of the line height from the bottom. 25 is 25% up from the bottom of the line, 100% is actually pretty much the same as the "filledBox" option).

| `brightBlue` | _Required_ | String | Sets the color used as ANSI bright blue. Uses hex color format: `"#rrggbb"`. |
| `brightCyan` | _Required_ | String | Sets the color used as ANSI bright cyan. Uses hex color format: `"#rrggbb"`. |
| `brightGreen` | _Required_ | String | Sets the color used as ANSI bright green. Uses hex color format: `"#rrggbb"`. |
| `brightPurple` | _Required_ | String | Sets the color used as ANSI bright purple. Uses hex color format: `"#rrggbb"`. |
| `brightRed` | _Required_ | String | Sets the color used as ANSI bright red. Uses hex color format: `"#rrggbb"`. |
| `brightWhite` | _Required_ | String | Sets the color used as ANSI bright white. Uses hex color format: `"#rrggbb"`. |
| `brightYellow` | _Required_ | String | Sets the color used as ANSI bright yellow. Uses hex color format: `"#rrggbb"`. |
| `cyan` | _Required_ | String | Sets the color used as ANSI cyan. Uses hex color format: `"#rrggbb"`. |
| `foreground` | _Required_ | String | Sets the foreground color of the color scheme. Uses hex color format: `"#rrggbb"`. |
| `green` | _Required_ | String | Sets the color used as ANSI green. Uses hex color format: `"#rrggbb"`. |

> | `name` | _Required_ | String | Name of the color scheme. |

Reviewer 4: I understand that you're alphabetizing the parameters by category (required, then optional)... but it is really awkward to have the `name` property for schemes be buried down here (as opposed to at the top) and to have the `foreground` and `background` properties not next to each other in the list.

Author: I'm happy to rearrange based on property relevance?

Reviewer 2: I'd probably do name, then fg/bg, then the rest of the colors
